File: lib/dotfiles/plan.rb
# frozen_string_literal: true

module Dotfiles
  # Collects and filters the state changes selected for an orchestration run.
  class Plan
    # @param actions [Array<Dotfiles::Action>] initial actions for the plan
    def initialize(actions = [])
      @actions = {}
      actions.each { |action| add(action) }
    end

    # Appends a state change while preserving the plan's fluent construction API.
    #
    # @param action [Dotfiles::Action]
    # @return [Dotfiles::Plan] this plan
    def add(action)
      raise "Duplicate action ID: #{action.id}" if @actions.key?(action.id)

      @actions[action.id] = action
      self
    end

    # @return [Array<Dotfiles::Action>] a read-only view of the planned actions
    def actions
      @actions.values.dup.freeze
    end

    # Reports whether this execution has any state changes to perform.
    #
    # @return [Boolean]
    def empty?
      @actions.empty?
    end

    # Reports how many state changes are selected.
    #
    # @return [Integer]
    def size
      @actions.size
    end

    # Selects shared and platform-specific state changes for the current host.
    #
    # @param platform [Symbol, String] friendly platform name
    # @return [Dotfiles::Plan]
    def for_platform(platform)
      selected_actions = @actions.values.select do |action|
        action.platform == :shared || action.platform.to_s == platform.to_s
      end

      self.class.new(selected_actions)
    end
  end
end

File: src/dotfiles/plan.py
"""Ordered action plans."""

from __future__ import annotations

from dotfiles.action import Action


class Plan:
    """Collect actions while preserving order and rejecting duplicate IDs."""

    def __init__(self, actions: list[Action] | None = None) -> None:
        self._actions: dict[str, Action] = {}

        for action in actions or []:
            self.add(action)

    def add(self, action: Action) -> Plan:
        """Add an action and return this plan for fluent construction."""
        if action.id in self._actions:
            raise ValueError(f"Duplicate action ID: {action.id}")

        self._actions[action.id] = action

        return self

    @property
    def actions(self) -> tuple[Action, ...]:
        """Return the ordered actions as an immutable view."""
        return tuple(self._actions.values())

    @property
    def empty(self) -> bool:
        """Return whether the plan contains no actions."""
        return not self._actions

    def for_platform(self, platform: str) -> Plan:
        """Return shared and platform-specific actions for the current host."""
        selected_actions = [
            action for action in self.actions if action.platform == "shared" or action.platform == platform
        ]

        return Plan(selected_actions)
